Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp

```diff
@@ -626,37 +666,184 @@ static std::pair<String, bool> cookiesForSession(const NetworkStorageSession& se
         }
     }
 
-    if (!cookies)
+    if (didAccessSecureCookies)
+        *didAccessSecureCookies = accessedSecureCookies;
+
+    return cookies;
+}
+
+std::pair<String, bool> NetworkStorageSession::lookupCookiesHeaders(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, ForHTTPHeader forHTTPHeader, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, IncludeSecureCookies includeSecureCookies) const
+{
+    bool didAccessSecureCookies = false;
+    auto cookies = lookupCookies(
+        firstParty,
+        sameSiteInfo,
+        url,
+        forHTTPHeader,
+        frameID,
+        pageID,
+        applyTrackingPrevention,
+        shouldRelaxThirdPartyCookieBlocking,
+        includeSecureCookies,
+        &didAccessSecureCookies);
+
+    if (!cookies || !*cookies)
         return { { }, false };
 
-    GUniquePtr<char> cookieHeader(soup_cookies_to_cookie_header(cookies));
-    soup_cookies_free(cookies);
+    GUniquePtr<char> cookieHeader(soup_cookies_to_cookie_header(*cookies));
+    soup_cookies_free(*cookies);
 
     return { String::fromUTF8(cookieHeader.get()), didAccessSecureCookies };
 }
 
-std::pair<String, bool> NetworkStorageSession::cookiesForDOM(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, IncludeSecureCookies includeSecureCookies, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking relaxThirdPartyCookieBlocking) const
+bool NetworkStorageSession::getRawCookies(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, Vector<Cookie>& rawCookies) const
+{
+    rawCookies.clear();
+
+    auto soupCookies = lookupCookies(
+        firstParty,
+        sameSiteInfo,
+        url,
+        ForHTTPHeader::No,
+        frameID,
+        pageID,
+        applyTrackingPrevention,
+        shouldRelaxThirdPartyCookieBlocking,
+        IncludeSecureCookies::Yes);
+
+    if (!soupCookies)
+        return false;
+
+    for (GSList* iter = *soupCookies; iter; iter = g_slist_next(iter)) {
+        GUniquePtr<SoupCookie> cookie(static_cast<SoupCookie*>(iter->data));
+        rawCookies.append(Cookie(cookie.get()));
+    }
+
+    g_slist_free(*soupCookies);
```

Comment:
This has been quite interesting to me -- I have learnt that `g_slist_free` won't release elements, and that works fine because the code has previously wrapped them into unique pointers that will take care of that -- and when a `Cookie` object is built, it copies things based on the raw cookie but won't keep any soon-to-be-invalid pointer.

Reply:
`Cookie::Cookie(SoupCookie* cookie)` copies all data and stores no references.

It is unclear but `GUniquePtr<SoupCookie> cookie(static_cast<SoupCookie*>(iter->data));` frees the cookie.

So the only data we have to manually free is the list.

Slightly awkward but I wanted to avoid doing something like making a Vector and converting back to GSList for libsoup, etc.

Reply:
Yes, the `GUniquePtr<SoupCookie>` part was fine -- I assumed it'd do the same as `unique_ptr` and would invoke some _destroy_ function on the `SoupCookie` as it went out of scope. 

What worried me was the `explicit Cookie(SoupCookie*)` constructor, as if it were to keep that pointer, it would be invalid, but it doesn't :+1: